treewide: Remove use of apple_sdk framework stub paths

booxter opened the pull request:
This patch covers all packages that use stub paths in some way in their
build instructions.

(Except those listed in [this](https://github.com/NixOS/nixpkgs/pull/399040#issuecomment-2814334550) comment.)

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

booxter commented:
I've manually checked these packages on aarch64-darwin. Results are as follows:

```
vengi-tools: already broken
cardpeek: already broken
gildas: builds
nextpnr: broken in staging-next
odin: builds
opensc: builds
bullet: builds
bullet-roboschool: already broken
libsamplerate: builds
netcdffortran: builds
jaxlib: builds
pyscard: builds
bazel-test-cpp: passes
xorg.xorgserver: already broken
```

I've also checked for those already broken packages that failures are similar to what can be seen on hydra. So I hope this doesn't break anything that wasn't broken before.

booxter commented:
Not sure if this commit / PR should have `treewide` as its title tag. Please advise.

NickCao commented:
The PR title should be treewide, but the changes should be in individual commits for individual packages.

booxter committed:
python3Packages.pyscard: clean up apple_sdk stubs

Signed-off-by: Ihar Hrachyshka <ihar.hrachyshka@gmail.com>

booxter committed:
netcdffortran: clean up apple_sdk stubs

Signed-off-by: Ihar Hrachyshka <ihar.hrachyshka@gmail.com>

booxter committed:
libsamplerate: clean up apple_sdk stubs

Signed-off-by: Ihar Hrachyshka <ihar.hrachyshka@gmail.com>

booxter committed:
xorg.xorgserver: clean up apple_sdk stubs

Signed-off-by: Ihar Hrachyshka <ihar.hrachyshka@gmail.com>

booxter committed:
odin: clean up apple_sdk stubs

Also, patched out all impure paths in the linker.cpp file for darwin.

Signed-off-by: Ihar Hrachyshka <ihar.hrachyshka@gmail.com>

booxter committed:
bullet{,roboschool}: clean up apple_sdk stubs

Signed-off-by: Ihar Hrachyshka <ihar.hrachyshka@gmail.com>

booxter committed:
gildas: clean up apple_sdk stubs

Signed-off-by: Ihar Hrachyshka <ihar.hrachyshka@gmail.com>

booxter committed:
opensc: clean up apple_sdk stubs

Signed-off-by: Ihar Hrachyshka <ihar.hrachyshka@gmail.com>

booxter committed:
bazel-test-cpp: clean up apple_sdk stubs

Signed-off-by: Ihar Hrachyshka <ihar.hrachyshka@gmail.com>

booxter committed:
python3Packages.jaxlib: clean up apple_sdk stubs

Note: jaxlib currently points to jaxlib-bin so this code is not utilized
for regular jaxlib installations.

Signed-off-by: Ihar Hrachyshka <ihar.hrachyshka@gmail.com>

booxter commented:
I separated each package into a separate commit. I also split out more elaborate pieces out into individual PRs:

- https://github.com/NixOS/nixpkgs/pull/399630
- https://github.com/NixOS/nixpkgs/pull/399609
- https://github.com/NixOS/nixpkgs/pull/399602
- https://github.com/NixOS/nixpkgs/pull/399600 (to fix build of a dependency for `xorg.xorgserver`)

With these merged, we should have no packages using old apple_sdk stubs in-tree.

Astavie commented:
How come this is being removed? I remember the odin package requiring these inputs to compile on macos, has there been a recent change so that is not required anymore?

yzx9 commented:
> How come this is being removed? I remember the odin package requiring these inputs to compile on macos, has there been a recent change so that is not required anymore?

Yes, the Darwin SDK has been fully reworked in #346043.
The old-style SDKs were just no-op placeholders now.

emilazy commented on pkgs/by-name/od/odin/package.nix:
```diff
@@ -24,17 +21,6 @@ stdenv.mkDerivation {
   };
 
-  postPatch =
-    lib.optionalString stdenv.hostPlatform.isDarwin ''
-      substituteInPlace src/linker.cpp \
-          --replace-fail '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk' ${MacOSX-SDK}
-    ''
```
Here’s the relevant piece of code: https://github.com/odin-lang/Odin/blob/4374685007fbf0398edfc8c94c39a8470572d9d0/src/linker.cpp#L772-L782.

This patching does currently have an effect; it replaces the `/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk` reference (problematic for building packages in Nix from a purity perspective) with an inert stub. However, the mentions of `/usr/local`, `/opt/homebrew`, and `/opt/local` seem comparably problematic. I suppose it is okay to leave the former unpatched since the latter already create impurities and the only Odin package we have in tree is `ols`, but I would personally be tempted to patch those lines out entirely.

booxter replied:
Correct. Updating.

emilazy commented on pkgs/by-name/op/opensc/package.nix:
```diff
@@ -35,43 +34,36 @@ stdenv.mkDerivation rec {
   PCSC_CFLAGS = lib.concatStringsSep " " (
     lib.optionals withApplePCSC [
-      "-I${darwin.apple_sdk.frameworks.PCSC}/Library/Frameworks/PCSC.framework/Headers"
       "-I${lib.getDev pcsclite}/include/PCSC"
```
Uh, do we really want to be using `pcsclite` headers when using Apple’s PCSC rather than `pcsclite`? I find this highly dubious; does it fail without?

emilazy commented on pkgs/development/libraries/bullet/default.nix:
```diff
@@ -51,10 +39,6 @@ stdenv.mkDerivation rec {
     ]
     ++ lib.optionals stdenv.hostPlatform.isDarwin [
       "-DOPENGL_FOUND=true"
```
The `FIND_PACKAGE` we were previously removing will set this; it should be dropped. Also I wouldn’t be surprised if we can remove this whole section and just let it build everything now.

emilazy commented on pkgs/development/libraries/bullet/roboschool-fork.nix:
```diff
@@ -52,10 +40,6 @@ stdenv.mkDerivation {
     ]
     ++ lib.optionals stdenv.hostPlatform.isDarwin [
       "-DOPENGL_FOUND=true"
```
(As with the other version of this package.)

emilazy reviewed (commented):
Looks great on the whole, thanks! Some nits.

smaret commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 399040 --package gildas`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>gildas</li>
  </ul>
</details>

smaret reviewed (approved):
The change in `gi/gildas/package.nix` looks good to me. Thanks!

emilazy commented on pkgs/by-name/od/odin/package.nix:
```diff
@@ -24,16 +21,12 @@ stdenv.mkDerivation {
-      patchShebangs build_odin.sh
-    '';
+  postPatch = lib.optionalString stdenv.hostPlatform.isDarwin ''
+    substituteInPlace src/linker.cpp \
+        --replace-fail '-Wl,-syslibroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -L/usr/local/lib ' "" \
+        --replace-fail '-L/opt/homebrew/lib ' "" \
+        --replace-fail '-L/opt/local/lib ' ""
```
nit: I would prefer a proper `.patch` that removes these lines entirely, to not have pointless runtime checks left in and to catch it at build time if they think of even more system paths to add.

emilazy commented on pkgs/development/python-modules/pyscard/default.nix:
```diff
@@ -30,9 +24,9 @@ buildPythonPackage rec {
+  nativeBuildInputs = [ swig ] ++ lib.optionals (!stdenv.hostPlatform.isDarwin) [ pkg-config ];
 
-  buildInputs = if withApplePCSC then [ PCSC ] else [ pcsclite ];
+  buildInputs = lib.optional (!stdenv.hostPlatform.isDarwin) pcsclite;
```
nit:
```suggestion
  buildInputs = lib.optionals (!stdenv.hostPlatform.isDarwin) [ pcsclite ];
```
for consistency (and because `lib.optional` is a slight footgun).

booxter replied:
@emilazy Will do; but could you please explain the footgun part so that I know why I'm doing it?

emilazy replied:
`lib.optional cond [ pkg ]` produces `[ [ pkg ] ]`, which doesn’t behave like you might expect. Since `lib.optionals` also scales better when you add more, I (and others) personally prefer using `lib.optionals` in all cases.

emilazy commented on pkgs/development/libraries/bullet/default.nix:
```diff
@@ -22,26 +20,16 @@ stdenv.mkDerivation rec {
-      Cocoa
-      OpenGL
-    ];
+  buildInputs = lib.optionals stdenv.hostPlatform.isLinux [
```
nit: `libGL`/`libGLU`/`libglut` are aliases that resolve to the correct thing for the platform, so this conditional can be removed, which would probably improve portability to e.g. FreeBSD.

emilazy commented on pkgs/development/libraries/bullet/roboschool-fork.nix:
```diff
@@ -26,36 +24,21 @@ stdenv.mkDerivation {
-      Cocoa
-      OpenGL
-    ];
+  buildInputs = lib.optionals stdenv.hostPlatform.isLinux [
```
nit: as before

emilazy reviewed (approved):
Looks great, thanks! I plan to do a `nixpkgs-review` and then merge within a day or two. Some nits if you feel like addressing them before then, but I won’t block on them.

emilazy reviewed (approved):
Okay, when I said I’d run `nixpkgs-review` I hadn’t looked at the rebuilds total. But I built every directly touched package on `x86_64-darwin` except for `xorg.xorgserver` (didn’t want to build on `staging` and too lazy to test it on top of the `libAppleWM` change, plus it was broken before anyway), `bullet-roboschool` (broken on `x86_64-darwin`), and `gildas` (just takes too long). Thank you!
